# Metagenomic_Transcriptomes
Identify what is in a metagenomic sample, and what it may be doing

## Purpose:
The advent of high-throughput sequencing methods has delivered a wealth of information to the scientific community, while simultaneously posing many new questions and creating significant data processing bottlenecks. One field high potential field for data-intensive applications is metagenomics, the culture-independent study of microbial genetic information from an environmental sample. Here we present an open-source, user-friendly software package for metagenomic functional analysis which bypasses taxonomic classification. The package utilizes existing software tools SRA tBLASTn, Trinity sequence assembly, RSEM transcript quantification, and BLASTx to create a streamlined pipeline for functional annotation and validation which can be used via the command line. Important features _____________. The pipeline was tested and validated using datasets ______________. We envision this software aiding to researchers seeking a functional understanding of metagenomic samples.

## Dependencies:
Python 2.7

## Directions for use:

## Example run command:

## Estimated runtime:


